### PR TITLE
ArC - remove unused interceptors

### PR DESCRIPTION
**Pull request by mkouba**
- and beans injected only by these interceptors
- and beans that are only injected in removed beans

### CODE REVIEW

**Comment by mkouba** [comment]
Things to add:

- [ ] test for "beans that are only injected in removed beans"
- [ ] remove unused decorators

**Comment by Sanne** [comment]
> * and beans injected only by these interceptors
> * and beans that are only injected in removed beans

Awesome!

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
Comment (Ladicek) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:368]: Copy 'n' paste :-)

**Comment by Sanne** [comment]
Looks good to me, but my knowledge of ArC is very superficial.

Also wondering: would you want to merge this now, or rather wait for 2.3+ ?

Regarding the algorythm: I guess it's a best effort, to be really sure to remove it all it should run in multiple phases until it neither any bean nor any interceptor is removed, no?

**Comment by Ladicek** [comment]
Looks fine on the first sight, though I'd have to check it out locally and walk around the code for a few hours to be confident :-)

@Sanne I think this method https://github.com/quarkusio/quarkus/blob/16fcd2295688bd699fc7dd48867db9ae124a3223/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java#L386-L421 does a fixed point iteration, which seems to be what you're asking for?

**Comment by Ladicek** [comment]
Ah I probably got it wrong. Indeed currently removing unused interceptors/decorators is only a single pass, which is followed by removing unused beans (which is iterated). I don't really know if it's possible for unused beans to declare interceptors that would become unused after the unused beans are removed, but I guess it might.

**Comment by Sanne** [comment]
Right, I think (if we wanted to really cover it all) something similar would need to be done while taking into account both beans and interceptors.

But I'm not sure if it's worth it - both in terms of complexity/maintenance and in terms of complexity/execution times. Was just bringing it up as I'm curious if it was a conscious choice.

**Comment by Ladicek** [comment]
OK, even better. So we remove unused beans at the beginning, then remove unused interceptors/decorators (that's single pass), then we remove unused beans again. I think it should be something like this instead (pseudocode):

```java
boolean shouldContinue = true;
while (shouldContinue) {
    boolean someBeansRemoved = removeUnusedBeans(...);
    boolean someInterceptorsRemoved = removeUnusedInterceptors(...);
    boolean someDecoratorsRemoved = removeUnusedDecorators(...);

    shouldContinue = someBeansRemoved || someInterceptorsRemoved || someDecoratorsRemoved;
}
```

**Review by Sanne — APPROVED** [review]
I'll approve so that @mkouba can merge if he wants to :+1: 

I'd be keen to measure it but my pet projects don't really use interceptors so I won't do this yet. I might do another pass to measure additional memory costs in `ArcContainerImpl`, but I do those in batches later.

**Comment by Sanne** [comment]
@Ladicek yes exactly that! Maybe you want to continue this after it's merged? up to @mkouba ;)

**Comment by mkouba** [comment]
> I don't really know if it's possible for unused beans to declare interceptors that would become unused after the unused beans are removed, but I guess it might.

Yes, I think that in theory it's possible. For example if you have a bean `Foo` which is only injected in an interceptor `Alpha` and it has an associated interceptor `Bravo` then currently we would only remove the unused interceptor `Alpha` and the unused bean `Foo`, but not the `Bravo` interceptor.

> Also wondering: would you want to merge this now, or rather wait for 2.3+ ?

That's a good question. It should probably target 2.3  because we may expect problems with programmatic lookup in extensions that do rely on less strict implementation (although the rules stay the same ;-).

> Regarding the algorithm: I guess it's a best effort, to be really sure to remove it all..

Yes, it's a best effort.  But we should probably try to cover maximum use cases...

CC @Sanne @Ladicek

**Comment by Ladicek** [comment]
Moving the fixed point iteration up one level shouldn't be hard, from my superficial understanding of the code. If @mkouba doesn't get to it, I may try later :-)

**Comment by mkouba** [comment]
I'm on it but I'll ping you for review ;-)

**Review by Ladicek — APPROVED** [review]
LGTM. My understanding of ArC internals is very shallow, so this is mostly a general approval, not rubber stamping of all the specific details :-)

**Comment by mkouba** [comment]
> LGTM. My understanding of ArC internals is very shallow, so this is mostly a general approval, not rubber stamping of all the specific details :-)

Don't be afraid I will not blame you if we find a critical bug in the relevant part of the code... or will I? ;-)

**Review by manovotn — APPROVED** [review]
This looks pretty awesome!

Out of curiosity, have you tried to measure the perf aspect of this PR with some quickstart app or custom app that uses interceptors/decorators? I don't expect much of a difference unless you start dealing with 500+ beans and their interceptors and decorators, but still curious to see if there is any real impact.

**Comment by mkouba** [comment]
> Out of curiosity, have you tried to measure the perf aspect of this PR with some quickstart app or custom app that uses interceptors/decorators? 

Not yet. In fact, any quarkus application has some built-in interceptors (`@ActivateRequestContext`, `@Lock` and `@CurrentThreadContext`).

> I don't expect much of a difference unless you start dealing with 500+ beans and their interceptors and decorators, but still curious to see if there is any real impact.

It shouldn't be a big difference. But note that this PR is not only about interceptors and decorators - we also remove beans that are only injected in removed beans and removed interceptors. And every next unused bean detected is a big win ;-). Imagine an `@ApplicationScoped` bean with 20 intercepted methods  - we need to generate a metadata class (i.e. impl of `Bean`), a client proxy class, a subclass and for every intercepted method a tiny class that implements a functional interface, i.e. we would save 23 classes, not to mention the classes that need to be loaded as dependencies.

**Comment by manovotn** [comment]
I understand, I was talking about a difference as in a perf cost for this detection. I have no doubts about gains of this PR :)

**Comment by mkouba** [comment]
> I understand, I was talking about a difference as in a perf cost for this detection

Well, I did a simple test with the `scheduler-quickstart`; executed 3x `mvn clean package -DskipTests -X` and got:

2.1.0.Final:
Removed 46 unused beans in `26`/`27`/`30` ms
Generated 39 resources in `96`/`101`/`109` ms

And this PR:
Removed 46 beans, 3 interceptors and 0 decorators in `23`/`30`/`34` ms
Generated 33 resources in `93`/`100`/`128` ms

**Review by geoand — APPROVED** [review]

**Comment by mkouba** [comment]
All CI jobs are green except for `BasicTest.shouldMakeJsonRequest()` from `quarkus-integration-test-resteasy-reactive-rest-clien` on JDK 11 Windows which failed with:
```
2021-07-28T11:13:53.5302870Z 2021-07-28 11:13:53,136 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (executor-thread-0) Client was closed, however the result was not completed. Handlers array is: [org.jboss.resteasy.reactive.client.handlers.ClientRequestFilterRestHandler, org.jboss.resteasy.reactive.client.handlers.ClientRequestFilterRestHandler, org.jboss.resteasy.reactive.client.handlers.ClientRequestFilterRestHandler, org.jboss.resteasy.reactive.client.handlers.ClientRequestFilterRestHandler, org.jboss.resteasy.reactive.client.handlers.ClientObservabilityHandler, org.jboss.resteasy.reactive.client.handlers.ClientSendRequestHandler, org.jboss.resteasy.reactive.client.handlers.ClientSetResponseEntityRestHandler, org.jboss.resteasy.reactive.client.handlers.ClientResponseFilterRestHandler, org.jboss.resteasy.reactive.client.handlers.ClientResponseFilterRestHandler, org.jboss.resteasy.reactive.client.handlers.ClientResponseCompleteRestHandler]. Last executed handler is: org.jboss.resteasy.reactive.client.handlers.ClientResponseCompleteRestHandler
2021-07-28T11:13:53.5321853Z 2021-07-28 11:13:53,143 INFO  [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-0) RoutingContext failure (500): javax.ws.rs.ProcessingException: Client request did not complete
2021-07-28T11:13:53.5324882Z 	at org.jboss.resteasy.reactive.client.impl.InvocationBuilderImpl.unwrap(InvocationBuilderImpl.java:210)
2021-07-28T11:13:53.5328187Z 	at org.jboss.resteasy.reactive.client.impl.InvocationBuilderImpl.method(InvocationBuilderImpl.java:326)
2021-07-28T11:13:53.5331366Z 	at io.quarkus.it.rest.client.main.AppleClient$$QuarkusRestClientInterface.swapApple(AppleClient$$QuarkusRestClientInterface.zig:633)
2021-07-28T11:13:53.5334471Z 	at io.quarkus.it.rest.client.main.ClientCallingResource.lambda$init$6(ClientCallingResource.java:67)
2021-07-28T11:13:53.5337166Z 	at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
2021-07-28T11:13:53.5339212Z 	at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:160)
2021-07-28T11:13:53.5340845Z 	at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:96)
2021-07-28T11:13:53.5510269Z 	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:158)
2021-07-28T11:13:53.5512454Z 	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
2021-07-28T11:13:53.5514322Z 	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:536)
2021-07-28T11:13:53.5516348Z 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2442)
2021-07-28T11:13:53.5518339Z 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1476)
2021-07-28T11:13:53.5520246Z 	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
2021-07-28T11:13:53.5523655Z 	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
2021-07-28T11:13:53.5526301Z 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
2021-07-28T11:13:53.5528007Z 	at java.base/java.lang.Thread.run(Thread.java:829)
2021-07-28T11:13:53.5529223Z Caused by: java.lang.IllegalStateException: Client request did not complete
2021-07-28T11:13:53.5531699Z 	at org.jboss.resteasy.reactive.client.impl.RestClientRequestContext.close(RestClientRequestContext.java:251)
2021-07-28T11:13:53.5536912Z 	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:187)
2021-07-28T11:13:53.5540953Z 	at org.jboss.resteasy.reactive.client.impl.AsyncInvokerImpl.performRequestInternal(AsyncInvokerImpl.java:260)
2021-07-28T11:13:53.5546289Z 	at org.jboss.resteasy.reactive.client.impl.AsyncInvokerImpl.performRequestInternal(AsyncInvokerImpl.java:251)
2021-07-28T11:13:53.5561137Z 	at org.jboss.resteasy.reactive.client.impl.AsyncInvokerImpl.method(AsyncInvokerImpl.java:211)
2021-07-28T11:13:53.5563011Z 	... 15 more
2021-07-28T11:13:53.5563424Z 
```
I don't think it's related so I'm going to rebase and squash all commits.

**Comment by geoand** [comment]
It is not related. It's some race condition that I yet to track down

**Comment by mkouba** [comment]
@Ladicek Just out of curiosity - have you ever seen this ^ `TimeoutUninterruptableTest.testTimeoutAsyncBulkhead()` TCK failure?

**Comment by Ladicek** [comment]
@mkouba Yea that's flaky (https://github.com/smallrye/smallrye-fault-tolerance/issues/233), but it seems to only occur in Quarkus :-) I've never seen it in SRye Fault Tolerance CI.

**Comment by mkouba** [comment]
Ok, so this PR is ready but should probably wait until the main branch corresponds to 2.3. @gsmet WDYT?

**Comment by mkouba** [comment]
> Ok, so this PR is ready but should probably wait until the main branch corresponds to 2.3. @gsmet WDYT?

OTOH according to the release planning this PR would have to wait at least few weeks, which is acceptable but may cause  headaches while resolving conflicts ;-).

**Review by gsmet — DISMISSED** [review]
Let's wait for 2.3. Please dismiss once 2.2 is merged.

**Review by gsmet — DISMISSED** [review]
Let's wait for 2.3. Please dismiss once 2.2 is branched.

**Pull request merged**
